libs/gl-client-py/src/credentials.rs

```diff
@@ -1,90 +1,145 @@
 use crate::runtime::exec;
 use crate::scheduler::Scheduler;
 use crate::signer::Signer;
-use crate::TlsConfig;
-use gl_client::credentials;
+use gl_client::credentials::{self, RuneProvider, TlsConfigProvider};
 use pyo3::exceptions::PyValueError;
 use pyo3::prelude::*;
 use pyo3::types::PyBytes;
 
-#[pyclass]
+pub type PyCredentials = InnerCredentials<credentials::Nobody, credentials::Device>;
+
 #[derive(Clone)]
-pub enum CredentialType {
-    Nobody = 0,
-    Device = 1,
+pub enum InnerCredentials<T, R>
+where
+    T: TlsConfigProvider,
+    R: TlsConfigProvider + RuneProvider + Clone,
+{
+    Nobody(T),
+    Device(R),
+}
+
+impl<T, R> InnerCredentials<T, R>
+where
+    T: TlsConfigProvider,
+    R: TlsConfigProvider + RuneProvider + Clone,
+{
+    pub fn is_nobody(&self) -> Result<()> {
+        if let Self::Nobody(_) = self {
+            Ok(())
+        } else {
+            Err(credentials::Error::IsIdentityError(
+                "credentials are not of type nobody".to_string(),
+            ))?
+        }
+    }
+
+    pub fn is_device(&self) -> Result<()> {
+        if let Self::Device(_) = self {
+            Ok(())
+        } else {
+            Err(credentials::Error::IsIdentityError(
+                "credentials are not of type device".to_string(),
+            ))?
+        }
+    }
```

Comment:
The `is_` prefix suggested to me that this was a `bool`-return, when really it is supposed to be used in combination with `?` to provide an abort criterium. We probably want to rename this to `ensure_device()` as it does just check.

Reply:
changed `is_` to `ensure_`